nvflare/private/aux_runner.py

```diff
@@ -141,6 +143,117 @@ def dispatch(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareab
                 valid_reply.set_cookie_jar(cookie_jar)
         return valid_reply
 
+    def _wait_for_cell(self):
+        start = time.time()
+        self.logger.debug("waiting for cell...")
+        max_wait = 5.0
+        while True:
+            cell = self.engine.get_cell()
+            if cell:
+                self.logger.debug(f"Got cell in {time.time() - start} secs")
+                return cell
+            if time.time() - start > max_wait:
+                self.logger.error(f"Cannot get cell after {max_wait} seconds!")
+                return None
+            time.sleep(0.01)
+
+    def _process_cell_replies(
+        self,
+        cell_replies: dict,
+        topic: str,
+        channel: str,
+    ):
+        replies = {}
+        if cell_replies:
+            for k, v in cell_replies.items():
+                assert isinstance(v, Message)
+                rc = v.get_header(MessageHeaderKey.RETURN_CODE, ReturnCode.OK)
+                target_name = FQCN.get_root(k)
+                if rc == CellReturnCode.OK:
+                    result = v.payload
+                    if not isinstance(result, Shareable):
+                        self.logger.error(f"reply of {channel}:{topic} must be Shareable but got {type(result)}")
+                        result = make_reply(ReturnCode.ERROR)
+                    replies[target_name] = result
+                else:
+                    src = self._convert_return_code(rc)
+                    replies[target_name] = make_reply(src)
+        return replies
+
+    def multicast_aux_requests(
+        self,
+        topic: str,
+        target_requests: Dict[str, Shareable],
+        timeout: float,
+        fl_ctx: FLContext,
+        optional: bool = False,
+        secure: bool = False,
+    ) -> dict:
+        if not target_requests:
+            return {}
+
+        # validate target names
+        target_names = [n for n in target_requests.keys()]
+        _, invalid_names = self.engine.validate_targets(target_names)
+        if invalid_names:
+            raise ValueError(f"invalid target(s): {invalid_names}")
+
+        try:
+            return self._send_multi_requests(
+                topic=topic,
+                target_requests=target_requests,
+                timeout=timeout,
+                fl_ctx=fl_ctx,
+                optional=optional,
+                secure=secure,
+            )
+        except Exception:
+            if optional:
```

Comment:
The flag of "optional" was introduced from 2.3. It means that this message is "nice to have" (e.g. last end-of-run message or getTask because it will be retried again) and it does not impact system functions even it failed. It is more than "debug".